src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace ILLink.RoslynAnalyzer
+{
+	[DiagnosticAnalyzer (LanguageNames.CSharp)]
+	public abstract class RequiresAnalyzerBase : DiagnosticAnalyzer
+	{
+		protected abstract string RequiresAttributeName { get; }
+
+		protected abstract string RequiresAttributeFullyQualifiedName { get; }
+
+		protected abstract DiagnosticTargets AnalyzerDiagnosticTargets { get; }
+
+		protected abstract DiagnosticDescriptor RequiresDiagnosticRule { get; }
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+			context.RegisterCompilationStartAction (context => {
+				var compilation = context.Compilation;
+				if (VerifyMSBuildOptions (context.Options, compilation))
+					return;
+				var dangerousPatterns = GetDangerousPatterns (compilation);
+
+				context.RegisterOperationAction (operationContext => {
+					var methodInvocation = (IInvocationOperation) operationContext.Operation;
+					CheckCalledMember (operationContext, methodInvocation.TargetMethod, dangerousPatterns);
+				}, OperationKind.Invocation);
+
+				context.RegisterOperationAction (operationContext => {
+					var objectCreation = (IObjectCreationOperation) operationContext.Operation;
+					var ctor = objectCreation.Constructor;
+					if (ctor is not null) {
+						CheckCalledMember (operationContext, ctor, dangerousPatterns);
+					}
+				}, OperationKind.ObjectCreation);
+
+				context.RegisterOperationAction (operationContext => {
+					var fieldAccess = (IFieldReferenceOperation) operationContext.Operation;
+					if (fieldAccess.Field.ContainingType is INamedTypeSymbol { StaticConstructors: var ctors } &&
+						!SymbolEqualityComparer.Default.Equals (operationContext.ContainingSymbol.ContainingType, fieldAccess.Field.ContainingType)) {
+						CheckStaticConstructors (operationContext, ctors);
```

Comment:
There is another call in CheckCalledMember that handles other IObjectCreationOperations

Reply:
I think we should do all-or-nothing here. Either no support for static constructors, or full support (including in the linker). That should probably go into another PR because it's not going to be a small change.

Reply:
Still need to remove all static constructor handling